File: internal/engine/hooks.go
package engine

// Lifecycle hooks: product-defined executables run at fixed points in the
// update sequence (docs/cli-contract.md). Each phase P runs every regular,
// executable file in <HooksDir>/P.d in lexical order, with update context
// in the environment (WENDY_*). Gating phases (pre-install, post-install,
// health) abort the update on the first non-zero exit; advisory phases
// (post-commit, on-failure) only log. Network-independent by design:
// products gate on local app/service readiness, not connectivity.

import (
	"bytes"
	"errors"
	"fmt"
	"log/slog"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"

	"github.com/wendylabsinc/wendy-os-update/internal/connector"
)

// DefaultHooksDir is the root holding the per-phase hook directories.
const DefaultHooksDir = "/etc/wendyos-update"

// Hook phases. The directory for a phase is <HooksDir>/<phase>.d.
const (
	HookPreInstall  = "pre-install"  // before writing the slot; non-zero aborts install
	HookPostInstall = "post-install" // after the swap, before reboot; non-zero aborts + unwinds
	HookHealth      = "health"       // commit gate, after platform verify; non-zero -> exit 4 -> rollback
	HookPostCommit  = "post-commit"  // after a successful commit; advisory (logged, never fatal)
	HookOnFailure   = "on-failure"   // a deployment was marked failed; advisory
)

// HookError reports a failing hook in a gating phase. cmd/wendyos-update maps
// it to an exit code (health -> 4, other gating phases -> 1).
type HookError struct {
	Phase string
	Hook  string
	Err   error
}

func (e *HookError) Error() string {
	return fmt.Sprintf("%s hook %q failed: %v", e.Phase, e.Hook, e.Err)
}

func (e *HookError) Unwrap() error { return e.Err }

// hookDir resolves a phase's hook directory. The health phase honours the
// legacy HealthDir override; every phase otherwise lives under HooksDir.
func (e *Engine) hookDir(phase string) string {
	if phase == HookHealth && e.HealthDir != "" {
		return e.HealthDir
	}
	root := e.HooksDir
	if root == "" {
		root = DefaultHooksDir
	}
	return filepath.Join(root, phase+".d")
}

// runHooks executes the phase's regular, executable files in lexical order,
// exporting WENDY_PHASE plus env to each. The first non-zero exit returns a
// *HookError. A missing or empty directory is a pass.
func (e *Engine) runHooks(phase string, env []string) error {
	dir := e.hookDir(phase)
	entries, err := os.ReadDir(dir)
	if os.IsNotExist(err) {
		return nil
	}
	if err != nil {
		return fmt.Errorf("read %s.d: %w", phase, err)
	}

	names := make([]string, 0, len(entries))
	for _, ent := range entries {
		if ent.IsDir() {
			continue
		}
		info, ierr := ent.Info()
		if ierr != nil || info.Mode()&0o111 == 0 {
			continue // skip non-executable files (READMEs, .conf, etc.)
		}
		names = append(names, ent.Name())
	}
	sort.Strings(names)

	if len(names) == 0 {
		slog.Debug("no hooks", "phase", phase, "dir", dir)
		return nil
	}

	hookEnv := append([]string{"WENDY_PHASE=" + phase}, env...)
	slog.Debug("hooks discovered", "phase", phase, "dir", dir,
		"count", len(names), "hooks", strings.Join(names, ","))
	slog.Debug("hook environment", "phase", phase, "env", strings.Join(hookEnv, " "))

	for _, name := range names {
		path := filepath.Join(dir, name)
		slog.Info("running hook", "phase", phase, "hook", name)
		cmd := exec.Command(path)
		// Tag the hook's own stdout/stderr with hook[<name>] and route it
		// through slog so it carries our journal severity prefix and is
		// greppable alongside the tool's output. Passing the SAME writer to
		// both streams makes os/exec serialize the two pipes' writes.
		hw := &hookLogWriter{name: name}
		cmd.Stdout = hw
		cmd.Stderr = hw
		cmd.Env = append(os.Environ(), hookEnv...)
		start := time.Now()
		err := cmd.Run()
		hw.flush()
		dur := time.Since(start).Round(time.Millisecond)
		if err != nil {
			slog.Error("hook failed", "phase", phase, "hook", name,
				"exit_code", exitCodeOf(err), "duration", dur, "err", err)
			return &HookError{Phase: phase, Hook: name, Err: err}
		}
		slog.Info("hook ok", "phase", phase, "hook", name, "duration", dur)
	}
	return nil
}

// exitCodeOf extracts the process exit code from a hook run error, or -1
// when the failure was not a non-zero exit (e.g. the binary could not be
// started — missing interpreter, not executable, permission denied).
func exitCodeOf(err error) int {
	var ee *exec.ExitError
	if errors.As(err, &ee) {
		return ee.ExitCode()
	}
	return -1
}

// hookLogWriter buffers a hook's output and emits each complete line through
// slog tagged with hook[<name>], so hook output is attributable and carries
// the same journal severity/prefix as the tool's own logs. Partial trailing
// output is held until the next newline or a final flush(); empty lines are
// dropped to keep the journal clean.
type hookLogWriter struct {
	name string
	buf  []byte
}

func (w *hookLogWriter) Write(p []byte) (int, error) {
	w.buf = append(w.buf, p...)
	for {
		i := bytes.IndexByte(w.buf, '\n')
		if i < 0 {
			break
		}
		w.emit(string(w.buf[:i]))
		w.buf = w.buf[i+1:]
	}
	return len(p), nil
}

func (w *hookLogWriter) flush() {
	if len(w.buf) > 0 {
		w.emit(string(w.buf))
		w.buf = nil
	}
}

func (w *hookLogWriter) emit(line string) {
	line = strings.TrimRight(line, "\r")
	if line == "" {
		return
	}
	slog.Info(fmt.Sprintf("hook[%s] %s", w.name, line))
}

// runAdvisoryHooks runs a non-gating phase: a failure is logged, never
// returned (post-commit, on-failure).
func (e *Engine) runAdvisoryHooks(phase string, env []string) {
	if err := e.runHooks(phase, env); err != nil {
		slog.Warn("advisory hook phase reported an error (ignored)", "phase", phase, "err", err)
	}
}

// hookEnv builds the update-context environment exposed to every hook
// (WENDY_PHASE is added per-phase by runHooks).
func (e *Engine) hookEnv(name, version string, target, cur connector.Slot, blUpdate bool) []string {
	return []string{
		"WENDY_ARTIFACT_NAME=" + name,
		"WENDY_ARTIFACT_VERSION=" + version,
		"WENDY_TARGET_SLOT=" + target.String(),
		"WENDY_CURRENT_SLOT=" + cur.String(),
		"WENDY_BOOTLOADER_UPDATE=" + strconv.FormatBool(blUpdate),
		"WENDY_STATE_DIR=" + e.StateDir,
	}
}

File: swift/Sources/Engine/Hooks.swift
import CLIError
import Connector
import Logging
import PlatformIO

// Lifecycle hooks: product-defined executables run at fixed points in the
// update sequence (docs/cli-contract.md). Each phase P runs every regular,
// executable file in <HooksDir>/P.d in lexical order, with update context
// in the environment (WENDY_*). Gating phases (pre-install, post-install,
// health) abort the update on the first non-zero exit; advisory phases
// (post-commit, on-failure) only log. Ports internal/engine/hooks.go.

private let logger = Logger(label: "wendyos-update")

/// Hook phases. The directory for a phase is `<HooksDir>/<phase>.d`.
public let HookPreInstall = "pre-install" // before writing the slot; non-zero aborts install
public let HookPostInstall = "post-install" // after the swap, before reboot; non-zero aborts + unwinds
public let HookHealth = "health" // commit gate, after platform verify; non-zero -> exit 4 -> rollback
public let HookPostCommit = "post-commit" // after a successful commit; advisory (logged, never fatal)
public let HookOnFailure = "on-failure" // a deployment was marked failed; advisory

/// Reports a failing hook in a gating phase. `WendyUpdate`'s top-level
/// error handling maps it to a process exit code via `ExitCoded` (health ->
/// 4, other gating phases -> 1). Ports `engine.HookError`; `underlying` is
/// this error's `Error.Error()` string (Go's `*exec.ExitError` renders as
/// `"exit status N"`) rather than a wrapped `Error` value, since Swift's
/// `Equatable` conformance needs a comparable payload and callers only ever
/// display this, never unwrap it (mirrors `HookError.Unwrap` having no
/// dedicated Swift caller either).
public struct HookError: Error, Equatable, ExitCoded {
    public let phase: String
    public let hook: String
    public let underlying: String

    public init(phase: String, hook: String, underlying: String) {
        self.phase = phase
        self.hook = hook
        self.underlying = underlying
    }

    public var exitCode: Int32 { phase == HookHealth ? 4 : 1 }
}

extension Engine {
    /// Resolves a phase's hook directory. The health phase honors the
    /// legacy `healthDir` override; every phase otherwise lives under
    /// `effectiveHooksDir`. Ports `Engine.hookDir`.
    func hookDir(_ phase: String) -> String {
        if phase == HookHealth, !healthDir.isEmpty {
            return healthDir
        }
        return Self.join(effectiveHooksDir, "\(phase).d")
    }

    /// Executes the phase's regular, executable files in lexical order,
    /// exporting `WENDY_PHASE` plus `env` to each. The first non-zero exit
    /// throws a `HookError`. A missing or empty directory is a pass. Ports
    /// `Engine.runHooks`.
    public func runHooks(_ phase: String, _ env: [String: String]) async throws {
        let dir = hookDir(phase)
        let entries: [DirEntry]
        do {
            entries = try fs.listDir(dir)
        } catch {
            // Missing directory is a pass — mirrors Go's
            // `os.IsNotExist(err) -> return nil`.
            return
        }

        let names = entries
            .filter { !$0.isDir && $0.isExecutable }
            .map(\.name)
            .sorted()

        if names.isEmpty {
            logger.debug("no hooks", metadata: ["phase": "\(phase)", "dir": "\(dir)"])
            return
        }

        var hookEnvVars = env
        hookEnvVars["WENDY_PHASE"] = phase

        logger.debug(
            "hooks discovered",
            metadata: [
                "phase": "\(phase)", "dir": "\(dir)",
                "count": "\(names.count)", "hooks": "\(names.joined(separator: ","))",
            ]
        )
        logger.debug(
            "hook environment",
            metadata: [
                "phase": "\(phase)",
                "env": "\(hookEnvVars.keys.sorted().map { "\($0)=\(hookEnvVars[$0]!)" }.joined(separator: " "))",
            ]
        )

        for name in names {
            let path = Self.join(dir, name)
            logger.info("running hook", metadata: ["phase": "\(phase)", "hook": "\(name)"])
            let exitCode: Int32
            do {
                exitCode = try await runner.runStreaming([path], env: hookEnvVars) { line in
                    logger.info("hook[\(name)] \(line)")
                }
            } catch {
                logger.error(
                    "hook failed",
                    metadata: ["phase": "\(phase)", "hook": "\(name)", "err": "\(error)"]
                )
                throw HookError(phase: phase, hook: name, underlying: "\(error)")
            }
            if exitCode != 0 {
                logger.error(
                    "hook failed",
                    metadata: ["phase": "\(phase)", "hook": "\(name)", "exit_code": "\(exitCode)"]
                )
                throw HookError(phase: phase, hook: name, underlying: "exit status \(exitCode)")
            }
            logger.info("hook ok", metadata: ["phase": "\(phase)", "hook": "\(name)"])
        }
    }

    /// Runs a non-gating phase: a failure is logged, never thrown
    /// (post-commit, on-failure). Ports `Engine.runAdvisoryHooks`.
    public func runAdvisoryHooks(_ phase: String, _ env: [String: String]) async {
        do {
            try await runHooks(phase, env)
        } catch {
            logger.warning(
                "advisory hook phase reported an error (ignored)",
                metadata: ["phase": "\(phase)", "err": "\(error)"]
            )
        }
    }

    /// Builds the update-context environment exposed to every hook
    /// (`WENDY_PHASE` is added per-phase by `runHooks`). Ports
    /// `Engine.hookEnv`.
    public func hookEnv(name: String, version: String, target: Slot, cur: Slot, blUpdate: Bool) -> [String: String] {
        [
            "WENDY_ARTIFACT_NAME": name,
            "WENDY_ARTIFACT_VERSION": version,
            "WENDY_TARGET_SLOT": target.description,
            "WENDY_CURRENT_SLOT": cur.description,
            "WENDY_BOOTLOADER_UPDATE": blUpdate ? "true" : "false",
            "WENDY_STATE_DIR": stateDir,
        ]
    }

    private static func join(_ dir: String, _ name: String) -> String {
        dir.hasSuffix("/") ? "\(dir)\(name)" : "\(dir)/\(name)"
    }
}
